File: js/srs.js
/* ==================================================================
   SRS — spaced repetition with real due dates.

   The previous "Practice Misses" mode weighted verses by lifetime
   accuracy: a verse you missed once a year ago and a verse you missed
   this morning looked identical, and a verse you had nailed three times
   running kept coming back at the same rate forever. Neither is how
   memory works.

   This is SM-2 (Anki's ancestor) with the standard easiness update,
   scheduling in whole days so that "due" means the same thing however
   the day is sliced up. Everything here is pure — no DOM, no clock
   beyond an injected `today` — because the scheduling is the part that
   has to be right and the part worth testing.
   ================================================================== */

var SRS = (function(){

  var MIN_EF = 1.3;
  var START_EF = 2.5;
  var MAX_INTERVAL = 365;

  /* Days since the epoch, in LOCAL time. Using UTC here would roll the
     day over mid-evening for anyone west of Greenwich and mid-morning
     for anyone far enough east — the daily streak would break for them
     while they were still awake on the same calendar day. */
  function dayNumber(date){
    var d = date || new Date();
    // Local calendar date, counted in UTC days. Taking the local Y/M/D and
    // reprojecting through Date.UTC keeps a daylight-saving shift from
    // moving the boundary by an hour and dropping or repeating a day.
    return Math.floor(Date.UTC(d.getFullYear(), d.getMonth(), d.getDate()) / 86400000);
  }

  function freshCard(){
    return { ef: START_EF, reps: 0, ivl: 0, due: 0, lapses: 0, last: 0 };
  }

  /* Answer quality on SM-2's 0-5 scale.
       5  right, and quickly — the recall was fluent
       4  right at a normal pace
       3  right but laboured, or propped up by a lifeline
       2  wrong, but the shape was there (a near miss when typing)
       1  wrong
       0  no answer at all
     `fraction` is time used as a share of the clock. */
  function gradeAnswer(o){
    o = o || {};
    if(o.timedOut) return 0;
    if(!o.correct) return o.near ? 2 : 1;
    if(o.usedPower) return 3;
    var f = typeof o.fraction === "number" ? o.fraction : 0.5;
    if(o.near) return 3;                 // typed it close enough, not exact
    if(f <= 0.4) return 5;
    if(f <= 0.8) return 4;
    return 3;
  }

  /* The SM-2 step. Returns a NEW card; never mutates the input. */
  function schedule(card, quality, today){
    var c = card ? {
      ef: card.ef, reps: card.reps, ivl: card.ivl,
      due: card.due, lapses: card.lapses, last: card.last
    } : freshCard();
    if(typeof c.ef !== "number" || !isFinite(c.ef)) c.ef = START_EF;
    var q = Math.max(0, Math.min(5, Math.round(quality)));

    if(q >= 3){
      if(c.reps === 0) c.ivl = 1;
      else if(c.reps === 1) c.ivl = 6;
      else c.ivl = Math.round(c.ivl * c.ef);
      c.reps += 1;
    } else {
      // A lapse sends the verse back to tomorrow but keeps its history,
      // so a verse you have repeatedly lost stays easy to spot.
      c.reps = 0;
      c.ivl = 1;
      c.lapses += 1;
    }

    // Standard SM-2 easiness update. Applied on every review, including
    // failures, which is what makes a chronically hard verse come back
    // faster than a merely new one.
    c.ef = c.ef + (0.1 - (5 - q) * (0.08 + (5 - q) * 0.02));
    if(c.ef < MIN_EF) c.ef = MIN_EF;

    if(c.ivl > MAX_INTERVAL) c.ivl = MAX_INTERVAL;
    c.last = today;
    c.due = today + c.ivl;
    return c;
  }

  function isDue(card, today){
    return !card || !card.reps ? true : card.due <= today;
  }
  /* Negative = not due yet. Bigger = more overdue. */
  function overdueBy(card, today){
    if(!card || !card.reps) return 0;
    return today - card.due;
  }

  /* Ordering for a drill. A session should open with what has actually
     fallen due, most overdue first, then bring in new material, and only
     then touch anything still comfortably scheduled.

       0  due (most overdue first)
       1  never seen
       2  seen, not yet due (soonest first) */
  function queueRank(card, today){
    if(!card || !card.reps){
      // A lapsed card has reps 0 but a history — that is not new material.
      if(card && card.lapses) return { band: 0, sort: -(today - card.due) };
      return { band: 1, sort: 0 };
    }
    if(card.due <= today) return { band: 0, sort: -(today - card.due) };
    return { band: 2, sort: card.due - today };
  }

  /* Build the drill list. `cardFor(verse)` returns that verse's card or
     undefined. Ties are broken by the caller-supplied shuffle so a drill
     is not identical every time. */
  function buildQueue(verses, cardFor, today, limit, shuffle){
    var mixed = (shuffle || function(a){ return a; })(verses.slice());
    var ranked = mixed.map(function(v, i){
      var r = queueRank(cardFor(v), today);
      return { v: v, band: r.band, sort: r.sort, i: i };
    });
    ranked.sort(function(a, b){
      return (a.band - b.band) || (a.sort - b.sort) || (a.i - b.i);
    });
    var out = ranked.map(function(x){ return x.v; });
    return typeof limit === "number" ? out.slice(0, limit) : out;
  }

  /* How many verses have actually fallen due. Never-seen verses are new
     material, not a backlog, so they are not counted — a first-time player
     should see "nothing due", not "305 due". A lapsed card still has to
     respect its due date like any other. */
  function dueCount(verses, cardFor, today){
    var n = 0;
    for(var i = 0; i < verses.length; i++){
      var c = cardFor(verses[i]);
      if(c && (c.reps || c.lapses) && c.due <= today) n++;
    }
    return n;
  }

  /* How settled a verse is, for the Study Hall. */
  function strength(card){
    if(!card || (!card.reps && !card.lapses)) return "unseen";
    if(!card.reps) return "lapsed";
    if(card.reps >= 3 && card.ivl >= 21) return "held";
    return "learning";
  }

  return {
    MIN_EF: MIN_EF, START_EF: START_EF, MAX_INTERVAL: MAX_INTERVAL,
    dayNumber: dayNumber, freshCard: freshCard, gradeAnswer: gradeAnswer,
    schedule: schedule, isDue: isDue, overdueBy: overdueBy, queueRank: queueRank,
    buildQueue: buildQueue, dueCount: dueCount, strength: strength
  };
})();

if(typeof module !== "undefined" && module.exports) module.exports = SRS;

File: mobile/core/src/main/kotlin/app/completetheverse/core/srs/Srs.kt
package app.completetheverse.core.srs

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.put
import java.time.Instant
import java.time.ZoneId
import kotlin.math.floor

data class SrsCard(
    val ef: Double = Srs.START_EF,
    val reps: Int = 0,
    val ivl: Int = 0,
    val due: Int = 0,
    val lapses: Int = 0,
    val last: Int = 0,
    val lastQuality: Int? = null,
    val lastMode: String? = null,
    val lastFraction: Double? = null,
    val lastCueLevel: Int? = null,
    val lastNear: Boolean? = null,
)

data class GradeInput(
    val timedOut: Boolean = false,
    val correct: Boolean = false,
    val near: Boolean = false,
    val usedPower: Boolean = false,
    val fraction: Double? = null,
    val cueLevel: Int = 0,
    val mode: String? = null,
)

data class QueueRank(
    val band: Int,
    val sort: Int,
)

object Srs {
    const val MIN_EF = 1.3
    const val START_EF = 2.5
    const val MAX_INTERVAL = 365

    fun dayNumber(
        date: Instant = Instant.now(),
        zone: ZoneId = ZoneId.systemDefault(),
    ): Int {
        // Local Y/M/D reprojected through UTC so DST cannot drop or repeat a day.
        val local = date.atZone(zone).toLocalDate()
        return local.toEpochDay().toInt()
    }

    fun freshCard(): SrsCard = SrsCard()

    fun gradeAnswer(o: GradeInput?): Int {
        val input = o ?: GradeInput()
        if (input.timedOut) return 0
        if (!input.correct) return if (input.near) 2 else 1
        if (input.usedPower) return 3
        val f = input.fraction ?: 0.5
        if (input.near) return 3
        if (f <= 0.4) return 5
        if (f <= 0.8) return 4
        return 3
    }

    fun schedule(card: SrsCard?, quality: Number, today: Int): SrsCard {
        val c = if (card != null) {
            card.copy(
                lastQuality = null,
                lastMode = null,
                lastFraction = null,
                lastCueLevel = null,
                lastNear = null,
            )
        } else {
            freshCard()
        }
        val ef = if (c.ef.isFinite()) c.ef else START_EF
        val q = jsRound(quality.toDouble()).coerceIn(0, 5)
        var reps = c.reps
        var ivl = c.ivl
        var lapses = c.lapses
        if (q >= 3) {
            ivl = when (reps) {
                0 -> 1
                1 -> 6
                else -> jsRound(ivl * ef)
            }
            reps += 1
        } else {
            reps = 0
            ivl = 1
            lapses += 1
        }
        var nextEf = ef + (0.1 - (5 - q) * (0.08 + (5 - q) * 0.02))
        if (nextEf < MIN_EF) nextEf = MIN_EF
        if (ivl > MAX_INTERVAL) ivl = MAX_INTERVAL
        return SrsCard(
            ef = nextEf,
            reps = reps,
            ivl = ivl,
            due = today + ivl,
            lapses = lapses,
            last = today,
        )
    }

    fun isDue(card: SrsCard?, today: Int): Boolean =
        if (card == null || card.reps == 0) true else card.due <= today

    fun overdueBy(card: SrsCard?, today: Int): Int {
        if (card == null || card.reps == 0) return 0
        return today - card.due
    }

    fun queueRank(card: SrsCard?, today: Int): QueueRank {
        if (card == null || card.reps == 0) {
            if (card != null && card.lapses != 0) {
                return QueueRank(band = 0, sort = -(today - card.due))
            }
            return QueueRank(band = 1, sort = 0)
        }
        if (card.due <= today) return QueueRank(band = 0, sort = -(today - card.due))
        return QueueRank(band = 2, sort = card.due - today)
    }

    fun <T> buildQueue(
        verses: List<T>,
        cardFor: (T) -> SrsCard?,
        today: Int,
        limit: Int? = null,
        shuffle: ((List<T>) -> List<T>)? = null,
    ): List<T> {
        val mixed = (shuffle ?: { it })(verses.toList())
        val ranked = mixed.mapIndexed { i, v ->
            val r = queueRank(cardFor(v), today)
            Ranked(v, r.band, r.sort, i)
        }
        val sorted = ranked.sortedWith(compareBy<Ranked<T>> { it.band }.thenBy { it.sort }.thenBy { it.i })
        val out = sorted.map { it.v }
        return if (limit != null) out.take(limit) else out
    }

    fun <T> dueCount(verses: List<T>, cardFor: (T) -> SrsCard?, today: Int): Int {
        var n = 0
        for (v in verses) {
            val c = cardFor(v)
            if (c != null && (c.reps != 0 || c.lapses != 0) && c.due <= today) n++
        }
        return n
    }

    fun strength(card: SrsCard?): String {
        if (card == null || (card.reps == 0 && card.lapses == 0)) return "unseen"
        if (card.reps == 0) return "lapsed"
        if (card.reps >= 3 && card.ivl >= 21) return "held"
        return "learning"
    }

    fun cardFromJson(el: JsonElement?): SrsCard? {
        val o = el as? JsonObject ?: return null
        return SrsCard(
            ef = jsonNumber(o["ef"], START_EF),
            reps = jsonNumber(o["reps"], 0.0).toInt(),
            ivl = jsonNumber(o["ivl"], 0.0).toInt(),
            due = jsonNumber(o["due"], 0.0).toInt(),
            lapses = jsonNumber(o["lapses"], 0.0).toInt(),
            last = jsonNumber(o["last"], 0.0).toInt(),
            lastQuality = o["lastQuality"]?.let { jsonNumber(it, 0.0).toInt() },
            lastMode = (o["lastMode"] as? JsonPrimitive)?.contentOrNull,
            lastFraction = o["lastFraction"]?.let { jsonNumber(it, Double.NaN) }?.takeIf { it.isFinite() },
            lastCueLevel = o["lastCueLevel"]?.let { jsonNumber(it, 0.0).toInt() },
            lastNear = o["lastNear"]?.let { jsonBool(it) },
        )
    }

    fun cardToJson(card: SrsCard): JsonObject = buildJsonObject {
        put("ef", card.ef)
        put("reps", card.reps)
        put("ivl", card.ivl)
        put("due", card.due)
        put("lapses", card.lapses)
        put("last", card.last)
        card.lastQuality?.let { put("lastQuality", it) }
        card.lastMode?.let { put("lastMode", it) }
        card.lastFraction?.let { put("lastFraction", it) }
        card.lastCueLevel?.let { put("lastCueLevel", it) }
        card.lastNear?.let { put("lastNear", it) }
    }

    fun cardsFromSave(srs: JsonElement?): Map<String, SrsCard> {
        val o = srs as? JsonObject ?: return emptyMap()
        val out = linkedMapOf<String, SrsCard>()
        for ((k, v) in o) {
            val card = cardFromJson(v) ?: continue
            out[k] = card
        }
        return out
    }

    fun putCard(save: JsonObject, verseId: String, card: SrsCard): JsonObject {
        val srs = ((save["srs"] as? JsonObject)?.toMutableMap() ?: mutableMapOf())
        srs[verseId] = cardToJson(card)
        val out = save.toMutableMap()
        out["srs"] = JsonObject(srs)
        return JsonObject(out)
    }

    private data class Ranked<T>(val v: T, val band: Int, val sort: Int, val i: Int)

    private fun jsonNumber(el: JsonElement?, fallback: Double): Double {
        if (el == null || el is JsonNull) return fallback
        val p = el as? JsonPrimitive ?: return fallback
        p.doubleOrNull?.let { return if (it.isNaN()) fallback else it }
        return p.content.toDoubleOrNull() ?: fallback
    }

    private fun jsonBool(el: JsonElement): Boolean {
        val p = el as? JsonPrimitive ?: return false
        return when (p.content) {
            "true" -> true
            "false" -> false
            else -> jsonNumber(p, 0.0) != 0.0
        }
    }
}

/** JS Math.round: half toward +Infinity for positives. */
internal fun jsRound(x: Double): Int {
    if (!x.isFinite()) return 0
    return floor(x + 0.5).toInt()
}
